streaming/src/main/scala/org/apache/spark/streaming/util/BatchedWriteAheadLog.scala

```diff
@@ -0,0 +1,191 @@
+  }
+
+  /**
+   * Delete the log files that are older than the threshold time.
+   *
+   * This method is handled by the parent WriteAheadLog.
+   */
+  override def clean(threshTime: Long, waitForCompletion: Boolean): Unit = {
+    parent.clean(threshTime, waitForCompletion)
+  }
+
+
+  /**
+   * Stop the manager, close any open log writer.
+   *
+   * This method is handled by the parent WriteAheadLog.
+   */
+  override def close(): Unit = {
+    writeAheadLogBatchWriter.stop()
+    batchWriterThreadPool.shutdownNow()
+    parent.close()
+  }
+
+  /** Start the actual log writer on a separate thread. */
+  private def startBatchedWriterThread(): BatchedLogWriter = {
+    val writer = new BatchedLogWriter()
+    val thread = new Thread(writer, "Batched WAL Writer")
+    thread.setDaemon(true)
+    thread.start()
+    writer
+  }
+
+  /** A helper class that writes LogEvents in a separate thread to allow for batching. */
+  private[util] class BatchedLogWriter extends Runnable {
+
+    private var active: Boolean = true
+    private val buffer = new ArrayBuffer[RecordBuffer]()
+
+    override def run(): Unit = {
+      while (active) {
+        try {
+          flushRecords()
+        } catch {
+          case NonFatal(e) =>
+            logError("Exception while flushing records in Batch Write Ahead Log writer.", e)
+        }
+      }
+      logInfo("Batch Write Ahead Log writer shutting down.")
+    }
+
+    def stop(): Unit = {
+      logInfo("Stopping Batch Write Ahead Log writer.")
+      active = false
+    }
+
+    /** Write all the records in the buffer to the write ahead log. */
+    private def flushRecords(): Unit = {
+      try {
+        buffer.append(walWriteQueue.take())
+        val numBatched = walWriteQueue.drainTo(buffer.asJava) + 1
+        logDebug(s"Received $numBatched records from queue")
+      } catch {
+        case _: InterruptedException =>
+          logWarning("Batch Write Ahead Log Writer queue interrupted.")
+      }
+      try {
+        var segment: WriteAheadLogRecordHandle = null
+        if (buffer.length > 0) {
+          logDebug(s"Batched ${buffer.length} records for Write Ahead Log write")
+          // we take the latest record for the time to ensure that we don't clean up files earlier
+          // than the expiration date of the records
+          val time = buffer.last.time
+          segment = parent.write(BatchedWriteAheadLog.aggregateRecords(buffer), time)
+        }
+        buffer.foreach(_.promise.success(segment))
+      } catch {
+        case NonFatal(e) =>
+          logWarning(s"Batch WAL Writer failed to write $buffer", e)
+          buffer.foreach(_.promise.success(null))
+      }
+      buffer.clear()
+    }
+  }
+}
+
+private[streaming] object BatchedWriteAheadLog {
+  private[streaming] def aggregateRecords(records: Seq[RecordBuffer]): ByteBuffer = {
+    ByteBuffer.wrap(Utils.serialize(
+      CombinedReceivedBlockTrackerLogEvent(records.map(_.record.array()).toArray)))
```

Comment:
Why is this tied with `CombinedReceivedBlockTrackerLogEvent`? This whole thing is completely inpdendent of the RBT.